osu.Game/Overlays/MusicController.cs

```diff
@@ -376,7 +376,7 @@ private bool next(bool allowProtectedTracks)
             {
                 Live<BeatmapSetInfo> result;
 
-                var possibleSets = getBeatmapSets().Where(s => !s.Value.Protected || allowProtectedTracks).ToArray();
+                var possibleSets = getBeatmapSets().Where(s => !s.Value.Equals(current?.BeatmapSetInfo) && (!s.Value.Protected || allowProtectedTracks)).ToArray();
```

Comment:
The test failure is the case where there's only a single track, and the expectation is that it should loop.

I'm not sure on the best way to resolve this. If the removal of the current track is moved further down, we could add an early return like:

```diff
diff --git a/osu.Game/Overlays/MusicController.cs b/osu.Game/Overlays/MusicController.cs
index 0f88765593..438b42b8ed 100644
--- a/osu.Game/Overlays/MusicController.cs
+++ b/osu.Game/Overlays/MusicController.cs
@@ -381,6 +381,9 @@ private bool next(bool allowProtectedTracks)
                 if (possibleSets.Length == 0)
                     return null;
 
+                if (possibleSets.Length == 1)
+                    return possibleSets.Single();
+
                 // condition below checks if the signs of `randomHistoryDirection` and `direction` are opposite and not zero.
                 // if that is the case, it means that the user had previously chosen next track `randomHistoryDirection` times and wants to go back,
                 // or that the user had previously chosen previous track `randomHistoryDirection` times and wants to go forward.

```

but i'm not sure how well this reads.

Reply:
Reads fine to me, so I've applied with a few extra inline comments.